Spring Data JPA - Method name parsing fails when fields start with "and" or "or"

Repositories method name parsing for does not handle fields starting with "and" or "or" and split by operators blindly.

With this repository: 
```java
public interface ProjectRepository extends Repository<Project, Long> {
    List<Project> findAllByOrganization(String organization);
}
```

Exception is thrown:
```
Caused by: io.quarkus.builder.BuildException:
Build failure: Build failed due to errors
	[error]: Build step io.quarkus.spring.data.deployment.SpringDataJPAProcessor#build threw an exception: io.quarkus.spring.data.deployment.UnableToParseMethodException: Entity org.acme.Project does not contain a field named: ganization. Offending method is findAllByOrganization
	at io.quarkus.spring.data.deployment.MethodNameParser.parse(MethodNameParser.java:210)
```

Simple project to reproduce it: 
```
git clone git@github.com:yall/quarkus-spring-data-parsing-bug.git
cd quarkus-spring-data-parsing-bug && ./mvnw test
```

--- a/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java
+++ b/extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/MethodNameParser.java
@@ -125,7 +125,7 @@ public class MethodNameParser {
 
         // handle the 'OrderBy' clause which is assumed to be at the end of the query
         Sort sort = null;
-        if (afterByPart.contains(ORDER_BY)) {
+        if (containsLogicOperator(afterByPart, ORDER_BY)) {
             int orderByIndex = afterByPart.indexOf(ORDER_BY);
             if (orderByIndex + ORDER_BY.length() == afterByPart.length()) {
                 throw new UnableToParseMethodException(
@@ -157,8 +157,8 @@ public class MethodNameParser {
         }
 
         List<String> parts = Collections.singletonList(afterByPart); // default when no 'And' or 'Or' exists
-        boolean containsAnd = afterByPart.contains("And");
-        boolean containsOr = afterByPart.contains("Or");
+        boolean containsAnd = containsLogicOperator(afterByPart, "And");
+        boolean containsOr = containsLogicOperator(afterByPart, "Or");
         if (containsAnd && containsOr) {
             throw new UnableToParseMethodException(
                     "'And' and 'Or' clauses cannot be mixed in a method name - Try specifying the Query with the @Query annotation. Offending method is "
@@ -364,6 +364,24 @@ public class MethodNameParser {
         String whereQuery = where.toString().isEmpty() ? "" : " WHERE " + where.toString();
         return new Result(entityClass, "FROM " + getEntityName() + whereQuery, queryType, paramsCount, sort,
                 topCount);
+    }
+
+    /**
+     * Meant to be called with {@param operator} being {@code "And"} or {@code "Or"}
+     * and returns {@code true} if the string contains the logical operator
+     * and the next character is an uppercase character.
+     * The reasoning is that if the next character is not uppercase,
+     * then the operator string is just part of a word
+     */
+    private boolean containsLogicOperator(String str, String operatorStr) {
+        int index = str.indexOf(operatorStr);
+        if (index == -1) {
+            return false;
+        }
+        if (str.length() < index + operatorStr.length() + 1) {
+            return false;
+        }
+        return Character.isUpperCase(str.charAt(index + operatorStr.length()));
     }
 
     /**

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/Post.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/Post.java
@@ -29,6 +29,8 @@ public class Post {
 
     private ZonedDateTime posted;
 
+    private String organization;
+
     public Long getId() {
         return id;
     }
@@ -67,6 +69,14 @@ public class Post {
 
     public void setPosted(ZonedDateTime postedAt) {
         this.posted = postedAt;
+    }
+
+    public String getOrganization() {
+        return organization;
+    }
+
+    public void setOrganization(String organization) {
+        this.organization = organization;
     }
 
     public void addComment(PostComment comment) {

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PostRepository.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PostRepository.java
@@ -13,4 +13,6 @@ public interface PostRepository extends JpaRepository<Post, Long> {
     Post findFirstByBypassTrue();
 
     List<Post> findByPostedBefore(ZonedDateTime zdt);
+
+    List<Post> findAllByOrganization(String organization);
 }

--- a/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PostResource.java
+++ b/integration-tests/spring-data-jpa/src/main/java/io/quarkus/it/spring/data/jpa/PostResource.java
@@ -42,6 +42,13 @@ public class PostResource {
 
     @GET
     @Produces("application/json")
+    @Path("/organization/{org}")
+    public List<Post> findAllByOrganization(@PathParam("org") String org) {
+        return postRepository.findAllByOrganization(org);
+    }
+
+    @GET
+    @Produces("application/json")
     @Path("/postComment/postId/{id}")
     public List<PostComment> findByPostId(@PathParam("id") Long id) {
         return postCommentRepository.findByPostId(id);

--- a/integration-tests/spring-data-jpa/src/main/resources/import.sql
+++ b/integration-tests/spring-data-jpa/src/main/resources/import.sql
@@ -34,9 +34,9 @@ INSERT INTO person(id, name, age, joined, active, address_id) VALUES (4, 'DeMar'
 INSERT INTO person(id, name, age, joined, active, address_id) VALUES (5, 'DeMar', 20, '2019-05-05', true, NULL);
 INSERT INTO person(id, name, age, joined, active, address_id) VALUES (6, null , 22, '2019-06-06', true, NULL);
 
-INSERT INTO post(id, title, bypass, posted) VALUES (1, 'Quarkus first public release!', false, '2019-03-01 12:00:00.000');
-INSERT INTO post(id, title, bypass, posted) VALUES (2, 'Quarkus 0.12.0 released', false, '2019-04-01 12:00:00.000');
-INSERT INTO post(id, title, bypass, posted) VALUES (3, 'Quarkus 0.20 released', false, '2019-06-01 12:00:00.000');
+INSERT INTO post(id, title, bypass, posted, organization) VALUES (1, 'Quarkus first public release!', false, '2019-03-01 12:00:00.000', 'RH');
+INSERT INTO post(id, title, bypass, posted, organization) VALUES (2, 'Quarkus 0.12.0 released', false, '2019-04-01 12:00:00.000', 'RH');
+INSERT INTO post(id, title, bypass, posted, organization) VALUES (3, 'Quarkus 0.20 released', false, '2019-06-01 12:00:00.000', 'RH');
 
 INSERT INTO post_comment(id, post_id, review) VALUES (1, 1, 'Excellent!');
 INSERT INTO post_comment(id, post_id, review) VALUES (2, 1, 'Wonderful!');

--- a/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/PostResourceTest.java
+++ b/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/PostResourceTest.java
@@ -41,6 +41,17 @@ public class PostResourceTest {
     }
 
     @Test
+    void testByByOrganization() {
+        when().get("/post/organization/RH").then()
+                .statusCode(200)
+                .body("size()", is(3));
+
+        when().get("/post/organization/RHT").then()
+                .statusCode(200)
+                .body("size()", is(0));
+    }
+
+    @Test
     void testPostCommentByPostId() {
         when().get("/post/postComment/postId/1").then()
                 .statusCode(200)

